### PR TITLE
[ITSASU-1158] Add throttling functionality

### PR DESCRIPTION
**Pull request by aug24**

### CODE REVIEW

File: test/controllers/utils/ReferenceRetrievalSpec.scala [hunk 1]
```diff
@@ -36,7 +36,8 @@ import services.mocks.MockSubscriptionDetailsService
 import services.{AuditingService, SubscriptionDetailsService}
 import uk.gov.hmrc.auth.core.{ConfidenceLevel, Enrolment, EnrolmentIdentifier, Enrolments}
 import uk.gov.hmrc.http.{HeaderCarrier, InternalServerException}
-import scala.concurrent.{Await, ExecutionContext, Future}
+
```
Comment (aug24) [inline test/controllers/utils/ReferenceRetrievalSpec.scala:39]: Just removing a warning

File: test/connectors/MockThrottlingConnector.scala [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+trait MockThrottlingConnector extends UnitTestTrait with MockitoSugar with BeforeAndAfterEach {
+
+  val mockThrottlingConnector: ThrottlingConnector = mock[ThrottlingConnector]
+
+  override def beforeEach(): Unit = {
+    super.beforeEach()
+    reset(mockThrottlingConnector)
+  }
+
+  def throttled(): OngoingStubbing[Future[Boolean]] = set(Future.successful(false))
```
Comment (aug24) [inline test/connectors/MockThrottlingConnector.scala:37]: Utility functions for use in mock testing.

File: test/services/ThrottlingServiceSpec.scala [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+      throttled()
+      val eventualResult = throttleable.testopen()
+      status(eventualResult) mustBe SEE_OTHER
+    }
+    "throttling throws an exception in a fail closed throttle" in {
+      throttleFail()
+      val eventualResult = throttleable.testclosed()
+      status(eventualResult) mustBe SEE_OTHER
+    }
+  }
```
Comment (sb-dev) [inline test/services/ThrottlingServiceSpec.scala:70]: A few suggestions to make it easier to understand:
```suggestion
  "throttled" should {
    "Return expected result" when {
      "throttling succeeds" in {
        val eventualResult = throttlingService.throttled(TestOpenThrottle) {
          Future.successful(NoContent)
        }

        Await.result(eventualResult, Duration(3, SECONDS)) mustBe NoContent
      }

      "throttling throws an exception in a fail open throttle" in {
        throttleFail()

        val eventualResult = throttlingService.throttled(TestOpenThrottle) {
          Future.successful(NoContent)
        }

        Await.result(eventualResult, Duration(3, SECONDS)) mustBe NoContent
      }
    }

    "Redirect as desired" when {
      "throttling fails" in {
        throttled()

        val eventualResult = throttlingService.throttled(TestOpenThrottle) {
          Future.successful(NoContent)
        }

        status(eventualResult) mustBe SEE_OTHER
      }

      "throttling throws an exception in a fail closed throttle" in {
        throttleFail()

        val eventualResult = throttlingService.throttled(TestClosedThrottle) {
          Future.successful(NoContent)
        }

        status(eventualResult) mustBe SEE_OTHER
      }
    }
  }
```

Comment (sb-dev) [inline test/services/ThrottlingServiceSpec.scala:70]: A few suggestions:
```suggestion
  "throttled" should {
    "Return expected result" when {
      "throttling succeeds" in {
        val eventualResult = throttlingService.throttled(TestOpenThrottle) {
          Future.successful(NoContent)
        }
        Await.result(eventualResult, Duration(3, SECONDS)) mustBe NoContent
      }
      "throttling throws an exception in a fail open throttle" in {
        throttleFail()
        val eventualResult = throttlingService.throttled(TestOpenThrottle) {
          Future.successful(NoContent)
        }
        Await.result(eventualResult, Duration(3, SECONDS)) mustBe NoContent
      }
    }
    "Redirect as desired" when {
      "throttling fails" in {
        throttled()
        val eventualResult = throttlingService.throttled(TestOpenThrottle) {
          Future.successful(NoContent)
        }
        status(eventualResult) mustBe SEE_OTHER
      }
      "throttling throws an exception in a fail closed throttle" in {
        throttleFail()
        val eventualResult = throttlingService.throttled(TestClosedThrottle) {
          Future.successful(NoContent)
        }
        status(eventualResult) mustBe SEE_OTHER
      }
    }
  }
```

**Review by sb-dev — APPROVED** [review]

**Pull request merged**